### PR TITLE
New: add fixer for "no-undef"

### PR DESCRIPTION
**Pull request by g-plane**

### CODE REVIEW

File: tests/lib/rules/no-undef.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    valid: [
+        "var test"
+    ],
+    invalid: [
+        {
+            code: "var test; text",
+            output: "var test; test",
+            errors: [{ messageId: "undef", type: "Identifier" }]
+        },
+        {
```
Comment (aladdin-add) [inline tests/lib/rules/no-undef.js]: what happens if the code is 
```js
var bat;
cat;
hat;
```

**Comment by aladdin-add** [comment]
seems like an interesting idea!

However, I have a concern it the autofix might change the runtime behavior, which the author don't expect.

File: tests/lib/rules/no-undef.js (hunk 1 above)
Comment (g-plane) [reply tests/lib/rules/no-undef.js]: It depends on the behavior of the library "string-similarity".

**Comment by g-plane** [comment]
Why do you think this will change runtime behavior? This rule only affects undeclared variables.

**Comment by aladdin-add** [comment]
e.g. `var test; foo(text)` => `var test; foo(test)`
if `text` is a global var, the changed code will have a different runtime behaviour.

**Comment by g-plane** [comment]
Seems reasonable. It's OK when using autofix in editor, however when running CLI the fix will be applied sliently.

**Comment by g-plane** [comment]
Hmm...Do you think fixers of other rules like `no-console` and `no-debugger` will change runtime behavior?

**Comment by aladdin-add** [comment]
Technically speaking, it did. But in most cases, users don't need to worry about it:

+ no-debugger: it has no effect if the user is not in dev mode.
+ no-console: it is safe unless something is watching the console output.

**Comment by g-plane** [comment]
Do you have any ideas? Or close this PR?

**Pull request closed without merging**